api/envoy/config/core/v3/protocol.proto

```diff
@@ -141,6 +141,10 @@ message UpstreamHttpProtocolOptions {
   // Does nothing if a filter before the http router filter sets the corresponding metadata.
   string override_auto_sni_header = 3
       [(validate.rules).string = {well_known_regex: HTTP_HEADER_NAME ignore_empty: true}];
+
+  // Alternative mechanism for derivation of the SNI value. It uses endpoint's :ref:`hostname <envoy_v3_api_field_config.endpoint.v3.Endpoint.hostname>` as the value, if ``hostname`` is set.
+  // If set, takes precedence over ``auto_sni``.
+  bool auto_sni_from_upstream = 4;
 }
```

Comment:
It's a minor point, but for consistency I think I would require auto_ani to be set as well, vs. just override it. Optimally this would have been a oneof but it's too late for that.

Reply:
Just a note: `oneof` is now discouraged by the API style guide as of https://github.com/envoyproxy/envoy/pull/30851

Reply:
@abeyad, @mattklein123, API style guide also discourages [boolean overloads](https://github.com/envoyproxy/envoy/blob/e100f231cc446b96a089e3f36d77181da6aa898f/api/STYLE.md?plain=1#L63) and advises usage of multiple fields with defined precedence.

Reply:
Alright sounds good.

Reply:
@dmitriyilin I think the way you have it set up now with precedence of `auto_sni_from_upstream` over `auto_sni` comports with my understanding of the API style guide.